pkg/static/login.js

```diff
@@ -747,15 +750,48 @@ import "./login.scss";
     function do_hostkey_verification(data) {
         const key_db = get_known_hosts_db();
         const key = data["host-key"];
-        const key_key = key.split(" ")[0];
-        const key_type = key.split(" ")[1];
+        const [key_key, key_type, key_value] = key.split(" ");
+        let db_key = key_db[key_key];
+
+        // did we see this host before?
+        if (db_key) {
+            // exact same key → good
+            if (key_db[key_key] == key) {
+                converse(data.id, data.default);
+                return;
+            }
 
-        if (key_db[key_key] == key) {
-            converse(data.id, data.default);
-            return;
+            /* transition case: cockpit-ssh stored full public keys in the DB, like
+            * "myhost ecdsa-sha2-nistp256 AAAA[..]ToM=\n"; ssh(1) does not show them, only fingerprints, so
+            * cockpit-beiboot only responds with fingerprints as well; they look like "myhost ED25519 SHA256:JVG[..]xs"
+            * check if we can match the given fingerprint to an existing entry */
+            const [db_key_type, db_key_value] = db_key.split(" ").slice(1);
+            if (key_value.startsWith('SHA256:') && db_key_value && !db_key_value.includes(':')) {
+                // cockpit-ssh and ssh use different notation: ED25519 vs. ssh-ed25519
+                if (key_type.replace('ssh-', '').toLowerCase() == db_key_type.replace('ssh-', '').toLowerCase()) {
+                    // cockpit-ssh and ssh agree no which key type to show to the user;
+                    // compute SHA256 fingerprint of reference db key and compare it
+                    const fp_raw = 'SHA256:' + base64_encode(sha256.digest(base64_decode(db_key_value)));
+                    // ssh omits trailing padding
+                    const fp = fp_raw.replace(/=+$/, '');
+
+                    // if it matches, accept
+                    if (fp == key_value) {
+                        converse(data.id, data.default);
+                        return;
+                    }
+
+                    // otherwise, the key changed, let user confirm below
+                } else {
+                    // cockpit-ssh and ssh disagree on which key type to show to the user
+                    console.warn("host key verification: host sent us key type", key_type,
+                                 "but our localStore database has type", db_key_type, "; treating as new host");
+                    db_key = null;
```

Comment:
I mentioned on IRC why I think this is a bad decision.

If someone wants to attack a user, and they know that this user uses Cockpit and recently went through this upgrade step then they can tune their attack by MITM-ing to a server which presents a key of a different type than the one that the user has in their DB in order that they get a more-innocent looking "new host" dialog instead of the "you are being attacked!!" one.  ie: in this scenario it's not a case of cockpit-ssh and ssh disagreeing, it's a case of the attacker having manipulated the scenario.

If we further assume that our attackers will always take advantage of this possibility to elicit a less-scary dialog, then it means that we'll actually only see the super-scary dialog in the case where we're *not* under attack (ie: key change for innocent reasons).

So ya: I don't like that we give the attacker the trivial ability to downgrade the type of warning.

Reply:
I think I misunderstood you. This situation can happen all the time, even with `ssh` itself. Over time and different server or libssh versions, the presented key type probably changed a lot. This PR doesn't fundamentally change that. But reading your full comment here I think that you merely want to show the "host key changed" dialog in this situation, which is fair (and I agree).

Reply:
Changed. This makes the patch much simpler actually, and it's more consistent. Sorry for the misunderstanding!